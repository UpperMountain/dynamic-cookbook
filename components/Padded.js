import React from "react";
import { View } from "react-native";
import { padding as padSize } from "../lib/theme";

const Padded = ({
  children,
  all,
  vertical,
  horizontal,
  left,
  right,
  top,
  bottom,
  style,
  ...rest
}) => {
  let _horizontal = horizontal;
  let _vertical = vertical;
  if (all) {
    _horizontal = true;
    _vertical = true;
  }
  const _top = (top || _vertical) * padSize;
  const _bottom = (bottom || _vertical) * padSize;
  const _left = (left || _horizontal) * padSize;
  const _right = (right || _horizontal) * padSize;

  return (
    <View
      style={{
        paddingLeft: _left,
        paddingRight: _right,
        paddingTop: _top,
        paddingBottom: _bottom,
        ...style
      }}
      {...rest}
    >
      {children}
    </View>
  );
};

export default Padded;
