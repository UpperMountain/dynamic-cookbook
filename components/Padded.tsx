import React from "react";
import { View, ViewProps } from "react-native";
import { padding as padSize } from "../lib/theme";

type Measure = boolean | number;
interface Props extends ViewProps {
  children: React.ReactNode;
  all?: Measure;
  vertical?: Measure;
  horizontal?: Measure;
  left?: Measure;
  right?: Measure;
  top?: Measure;
  bottom?: Measure;
  style?: ViewProps["style"];
}

const Padded = ({
  children,
  all,
  vertical,
  horizontal,
  left,
  right,
  top,
  bottom,
  style,
  ...rest
}: Props) => {
  let _horizontal = horizontal;
  let _vertical = vertical;
  if (all) {
    _horizontal = true;
    _vertical = true;
  }
  const _top = Number(top || _vertical || 0) * padSize;
  const _bottom = Number(bottom || _vertical || 0) * padSize;
  const _left = Number(left || _horizontal || 0) * padSize;
  const _right = Number(right || _horizontal || 0) * padSize;

  return (
    <View
      style={[
        {
          paddingLeft: _left,
          paddingRight: _right,
          paddingTop: _top,
          paddingBottom: _bottom
        },
        style
      ]}
      {...rest}
    >
      {children}
    </View>
  );
};

export default Padded;
