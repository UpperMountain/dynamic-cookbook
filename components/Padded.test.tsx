import React from "react";
import { Text } from "react-native";
import Padded from "./Padded";
import renderer from "react-test-renderer";

it("should render without exploding", () => {
  renderer
    .create(
      <Padded all>
        <Text>testing</Text>
      </Padded>
    )
    .unmount();
});

it("should pass snapshow tests", () => {
  const cmp = renderer.create(
    <>
      <Padded top>
        <Text>testing</Text>
      </Padded>
      <Padded bottom>
        <Text>testing</Text>
      </Padded>
      <Padded left>
        <Text>testing</Text>
      </Padded>
      <Padded right>
        <Text>testing</Text>
      </Padded>
      <Padded vertical>
        <Text>testing</Text>
      </Padded>
      <Padded horizontal>
        <Text>testing</Text>
      </Padded>
      <Padded all>
        <Text>testing</Text>
      </Padded>
    </>
  );
  expect(cmp.toJSON()).toMatchSnapshot();
  cmp.unmount();
});

it("should consistently handle multiple parameters", () => {
  const cmp = renderer.create(
    <>
      <Padded top={1} left={0.5}>
        <Text>testing</Text>
      </Padded>
      <Padded bottom={2} horizontal={3}>
        <Text>testing</Text>
      </Padded>
      <Padded all={3} left={10}>
        <Text>testing</Text>
      </Padded>
      <Padded left={2} bottom={1}>
        <Text>testing</Text>
      </Padded>
    </>
  );
  expect(cmp.toJSON()).toMatchSnapshot();
  cmp.unmount();
});
