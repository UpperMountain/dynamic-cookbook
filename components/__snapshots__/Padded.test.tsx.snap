// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should consistently handle multiple parameters 1`] = `
Array [
  <View
    style={
      Array [
        Object {
          "paddingBottom": 0,
          "paddingLeft": 10,
          "paddingRight": 0,
          "paddingTop": 20,
        },
        undefined,
      ]
    }
  >
    <Text>
      testing
    </Text>
  </View>,
  <View
    style={
      Array [
        Object {
          "paddingBottom": 40,
          "paddingLeft": 60,
          "paddingRight": 60,
          "paddingTop": 0,
        },
        undefined,
      ]
    }
  >
    <Text>
      testing
    </Text>
  </View>,
  <View
    style={
      Array [
        Object {
          "paddingBottom": 20,
          "paddingLeft": 200,
          "paddingRight": 20,
          "paddingTop": 20,
        },
        undefined,
      ]
    }
  >
    <Text>
      testing
    </Text>
  </View>,
  <View
    style={
      Array [
        Object {
          "paddingBottom": 20,
          "paddingLeft": 40,
          "paddingRight": 0,
          "paddingTop": 0,
        },
        undefined,
      ]
    }
  >
    <Text>
      testing
    </Text>
  </View>,
]
`;

exports[`should pass snapshow tests 1`] = `
Array [
  <View
    style={
      Array [
        Object {
          "paddingBottom": 0,
          "paddingLeft": 0,
          "paddingRight": 0,
          "paddingTop": 20,
        },
        undefined,
      ]
    }
  >
    <Text>
      testing
    </Text>
  </View>,
  <View
    style={
      Array [
        Object {
          "paddingBottom": 20,
          "paddingLeft": 0,
          "paddingRight": 0,
          "paddingTop": 0,
        },
        undefined,
      ]
    }
  >
    <Text>
      testing
    </Text>
  </View>,
  <View
    style={
      Array [
        Object {
          "paddingBottom": 0,
          "paddingLeft": 20,
          "paddingRight": 0,
          "paddingTop": 0,
        },
        undefined,
      ]
    }
  >
    <Text>
      testing
    </Text>
  </View>,
  <View
    style={
      Array [
        Object {
          "paddingBottom": 0,
          "paddingLeft": 0,
          "paddingRight": 20,
          "paddingTop": 0,
        },
        undefined,
      ]
    }
  >
    <Text>
      testing
    </Text>
  </View>,
  <View
    style={
      Array [
        Object {
          "paddingBottom": 20,
          "paddingLeft": 0,
          "paddingRight": 0,
          "paddingTop": 20,
        },
        undefined,
      ]
    }
  >
    <Text>
      testing
    </Text>
  </View>,
  <View
    style={
      Array [
        Object {
          "paddingBottom": 0,
          "paddingLeft": 20,
          "paddingRight": 20,
          "paddingTop": 0,
        },
        undefined,
      ]
    }
  >
    <Text>
      testing
    </Text>
  </View>,
  <View
    style={
      Array [
        Object {
          "paddingBottom": 20,
          "paddingLeft": 20,
          "paddingRight": 20,
          "paddingTop": 20,
        },
        undefined,
      ]
    }
  >
    <Text>
      testing
    </Text>
  </View>,
]
`;
